compiler/noirc_evaluator/src/ssa/opt/constant_folding_brillig_calls.rs

```diff
@@ -0,0 +1,526 @@
+        function::Function,
+        instruction::{Instruction, InstructionId},
+        types::Type,
+        value::{Value as IrValue, ValueId},
+    },
+    ssa_gen::Ssa,
+};
+
+impl Ssa {
+    /// See [`constant_folding_brillig_calls`][self] module for more information.
+    pub(crate) fn fold_constant_brillig_calls(mut self) -> Ssa {
+        // Gather constant evaluation results per function
+        let mut constant_evaluations = self
+            .functions
+            .iter()
+            .map(|(id, function)| (*id, function.constant_evaluation(&self)))
+            .collect::<HashMap<_, _>>();
+
+        // Apply constant folding for each function based on the interpreter results
+        for function in self.functions.values_mut() {
+            let Some(constant_evaluations) = constant_evaluations.remove(&function.id()) else {
+                continue;
+            };
+
+            function.fold_constant_evaluation(constant_evaluations);
+        }
+
+        self
+    }
+}
+
+impl Function {
+    /// For a given function, identifies all Brillig function calls with constant arguments,
+    /// interprets them using the [SSA interpreter][crate::ssa::interpreter], and records the result.
+    ///
+    /// This does not mutate the IR. The interpreter borrows the entire Ssa object so we split up
+    /// the actual interpreter execution from the mutation of the IR.
+    fn constant_evaluation(&self, ssa: &Ssa) -> HashMap<InstructionId, Vec<Value>> {
+        let mut instr_to_const_result = HashMap::default();
+        // We must track the constant results of Brillig calls in case those results
+        // are used as the input to another Brillig call.
+        let mut const_result_values: HashMap<ValueId, Value> = HashMap::default();
+
+        // Only ACIR functions can be evaluated in this pass.
+        // Brillig non-entry points themselves are not subject to constant evaluation.
+        if self.runtime().is_brillig() {
+            return instr_to_const_result;
+        }
+
+        for block_id in self.reachable_blocks() {
+            for instruction_id in self.dfg[block_id].instructions() {
+                let instruction = &self.dfg[*instruction_id];
+                let Instruction::Call { func: func_id, arguments } = instruction else {
+                    continue;
+                };
+
+                let func_value = &self.dfg[*func_id];
+                let IrValue::Function(func_id) = func_value else {
+                    continue;
+                };
+
+                let Some(func) = ssa.functions.get(func_id) else {
+                    unreachable!("ICE: Function {func_id} does not exist, yet it is being called");
+                };
+
+                // Skip calls to non-Brillig functions
+                if !func.runtime().is_brillig() {
+                    continue;
+                }
+
+                let mut all_constants = true;
+                let mut interpreter_args = Vec::new();
```

Comment:
Not sure how it affects performance but I wonder if splitting this into two things:
1. check if all arguments are constant
2. creating the `interpreter_args` array
would be better. For example if we have a brillig call with a large constant array as the first argument but a dynamic second argument we'd go through `const_ir_value_to_interpreter_value`, etc.
This is a minor thing, though.

Reply:
done